src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1619,32 +1618,35 @@ private RefreshItem getRefreshToken(final AuthenticationRequest request) {
         final String methodName = ":getRefreshToken";
         RefreshItem refreshItem = null;
         if (mTokenCacheStore != null) {
-            boolean multiResource = false;
-            // target refreshToken for this resource first. CacheKey will
-            // include the resourceId in the cachekey
-            Logger.v(TAG, "Looking for regular refresh token");
             String userId = request.getUserId();
             if (StringExtensions.IsNullOrBlank(userId)) {
                 // acquireTokenSilent expects userid field from UserInfo
                 userId = request.getLoginHint();
             }
-            String keyUsed = CacheKey.createCacheKey(request, userId);
+
+            // Target multiResource item in cache first because it's the most recent one.
+            // Cache key will not include resourceId.
+            Logger.v(TAG, "Looking for Multi Resource Refresh token");
+            String keyUsed  = CacheKey.createMultiResourceRefreshTokenKey(request, userId);
```

Comment:
See https://github.com/AzureAD/azure-activedirectory-library-for-android/blob/69198ef1317fd15b473c3300c74ac8453a5a14ee/src/src/com/microsoft/aad/adal/Oauth2.java#L241 for how to determine whether a returned RT is in fact an MRRT.  At this time all RTs returned from AAD are MRRTs.  ADAL doesn't have a way of determining whether it is actually talking to AAD unfortunately, and in the future AADs behavior may change.  Thus, there is a precise way of determining whether the returned RT is in fact an MRRT or not.  Even more unfortunately, the method is of determining whether an RT is in fact an MRRT is absolutely not intuitive.  If the response from the server includes a specific resource then it is in fact an MRRT.  If it does not, then it is not an MRRT.

It appears that the first time the Android library receives an MRRT it treats it as special and places it in its own entry, and then puts a copy into the entry with the AT.  This is broken.  It would have been far better if it had stored exactly one copy of the MRRT.  None of the AT entries should have an RT.  They should instead refer directly to the MRRT.  Each time a new AT is needed the MRRT should be used.  Then the current MRRT should be replaced with the one that was received along with the AT.  Unfortunately, that was made difficult by a poor cache extension model that we were working with in the beginning.  The alternative approach is to not have a separate entry for the MRRT at all.  Instead, each AT is stored with the most current MRRT.  Whenever an MRRT is used, and a new MRRT is returned, then all of the existing copies need to be updated with the new one.  This is how the node.js ADAL works.

NOTE: The current FoCI design is completely contradictory to this and that is probably causing confusion.  However, the FoCI design is being changed so that it will in fact be symmetrical to this design.

The current Android library did not use either of these methods, and that is a major bug that you have discovered.  Whatever we change, we need to retain backward compatibility for anyone that may have decided to extend the cache.  Thus I am suggesting that you switch the code to use the first algorithm I detailed above, namely having a single master copy of the MRRT, but save a copy with the current AT for older consumers that expect the RT to be there as well.

So to go in more depth on your questions:

1) We should know whether an AT was returned along with an MRRT because we should be checking that whenever we get an AT and RT.  If we are not checking that in every case it is a bug.  We should also be saving that fact in every AT entry.  Whenever we need a AT we should first check if the AT was returned along with an MRRT, and if it was then we should be using the master copy of the MRRT.
2) The regular RT you refer to may not actually be a regular RT.  In fact, the only time that it would be is if the RT was obtained directly from ADFS rather than AAD.  The only way to know that for sure is to check and record whether the AT was returned along with an MRRT every single time we get a server response.
3) I think this question is essentially answered by the above.
4) As above, if an AT was ever returned with an MRRT then all subsequent AT refreshes will be returned with an MRRT as well.
5) This is why I am suggesting that the logic should be to always use the single master MRRT but save refreshed MRRT along with the AT anyway.  The code should no longer ever use the RT/MRRT in the AT entry if we know that the AT was returned with an MRRT.